core/corehttp/ipns_hostname.go

```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```

Comment:
In #1233 the last left problem is not what your thesis states.

The last left problem is we are taking the DNS entry `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/` (dont miss the trailing slash!) and put the request path (`/`) on it. So we have `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA//`which is redirected by the mux (not by any IPFS code) to `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/`. After that we take the DNS entry and attach the path so we have `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA//ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA/` which contains a `//` and again the mux is intercepting.

- So one solution would just be ensure there are no `//` in the path after attaching the path to it. (Done that with `strings.TrimRight`)
- The other solution is to use IPNS. (Explained above)

Regarding domains:
- My public gateway (ipfs.hub.luzifer.io) has no `dnslink` TXT entry so this does not apply to those gateways.
- swagger.luzifer.io is a website. Not a gateway. If someone tries to use it as a gateway it should fail by design. Accessing content outside `/ipfs/QmczrnBpAgnhCo87rtZ8VbLfGU5v8g7cKr7Ef2fKRvYniA` would be gateway functionality…